docs/tutorial/part_2.rst

```diff
@@ -0,0 +1,459 @@
+
+.. note::
+    It is good practice to use a common path prefix like ``/ws/`` to distinguish
+    WebSocket connections from ordinary HTTP connections because it will make
+    deploying Channels to a production environment in certain configurations
+    easier.
+    
+    In particular for large sites it will be possible to configure a
+    production-grade HTTP server like nginx to route requests based on path to
+    either (1) a production-grade WSGI server like Gunicorn+Django for ordinary
+    HTTP requests or (2) a production-grade ASGI server like Daphne+Channels
+    for WebSocket requests.
+    
+    Note that for smaller sites you can use a simpler deployment strategy where
+    Daphne serves all requests - HTTP and WebSocket - rather than having a
+    separate WSGI server. In this deployment configuration no common path prefix
+    like is ``/ws/`` is necessary.
+
+Create a new file **chat/consumers.py**. Your app directory should now look like::
+
+    chat/
+        __init__.py
+        consumers.py
+        templates/
+            chat/
+                index.html
+                room.html
+        urls.py
+        views.py
+
+Put the following code in **chat/consumers.py**::
+
+    # chat/consumers.py
+    from channels.generic.websocket import WebsocketConsumer
+    import json
+    
+    class ChatConsumer(WebsocketConsumer):
+        def connect(self):
+            self.accept()
+        
+        def disconnect(self, close_code):
+            pass
+        
+        def receive(self, text_data):
+            text_data_json = json.loads(text_data)
+            message = text_data_json['message']
+            
+            self.send(text_data=json.dumps({
+                'message': message
+            }))
+
+This is a synchronous WebSocket consumer that accepts all connections, receives
+messages from its client, and echos those messages back to the same client. For
+now it does not broadcast messages to other clients in the same room.
+
+.. note::
+    Channels also supports writing *asynchronous* consumers for greater
+    performance. However any asynchronous consumer must be careful to avoid
+    directly performing blocking operations, such as accessing a Django model.
+    See the :doc:`/topics/consumers` reference for more information about writing asynchronous
+    consumers.
+
+We need to create a routing configuration for the **chat** app that has a route to
+the consumer. Create a new file **chat/routing.py**. Your app directory should now
+look like::
+
+    chat/
+        __init__.py
+        consumers.py
+        routing.py
+        templates/
+            chat/
+                index.html
+                room.html
+        urls.py
+        views.py
+
+Put the following code in **chat/routing.py**::
+
+    # chat/routing.py
+    from django.conf.urls import url
+    
+    from . import consumers
+    
+    websocket_urlpatterns = [
+        url("^ws/chat/(?P<room_name>[^/]+)/$", consumers.ChatConsumer),
```

Comment:
While not strictly required in this case, I'd recommend making this a `r""` string.

Reply:
Nice catch!